include/mbedtls/ssl.h

```diff
@@ -4891,6 +4884,151 @@ int mbedtls_ssl_send_alert_message( mbedtls_ssl_context *ssl,
+ *                 specification of mbedtls_ssl_write() relevant to TLS 1.3
+ *                 (thus not the parts specific to (D)TLS1.2) applies to this
+ *                 function and the present documentation is restricted to the
+ *                 differences with mbedtls_ssl_write().
+ *
+ * \param ssl      SSL context
+ * \param buf      buffer holding the data
+ * \param len      how many bytes must be written
+ *
+ * \return         One additional specific return value:
+ *                 #MBEDTLS_ERR_SSL_CANNOT_WRITE_EARLY_DATA.
+ *
+ *                 #MBEDTLS_ERR_SSL_CANNOT_WRITE_EARLY_DATA is returned when it
+ *                 is not possible to write early data for the SSL context
+ *                 \p ssl.
+ *
+ *                 It may have been possible and it is not possible
+ *                 anymore because the client received the server Finished
+ *                 message, the server rejected early data or the maximum
+ *                 number of allowed early data for the PSK in use has been
+ *                 reached.
+ *
+ *                 It may never have been possible and will never be possible
+ *                 for the SSL context \p ssl because the use of early data
+ *                 is disabled for that context or more generally the context
+ *                 is not suitably configured to enable early data or the first
+ *                 call to the function was done while the handshake was
+ *                 already completed.
+ *
+ *                 It is not possible to write early data for the SSL context
+ *                 \p ssl but this does not preclude for using it with
+ *                 mbedtls_ssl_write(), mbedtls_ssl_read() or
+ *                 mbedtls_ssl_handshake().
+ *
+ * \note           This function may write early data only if the SSL context
+ *                 has been configured for the handshake with a PSK for which
+ *                 early data is allowed.
+ *
+ * \note           To maximize the number of early data that can be written in
+ *                 the course of the handshake, it is expected that this
+ *                 function starts the handshake for the SSL context \p ssl.
+ *                 But this is not mandatory.
+ *
+ * \note           This function does not provide any information on whether
+ *                 the server has accepted or will accept early data or not.
+ *                 When it returns a positive value, it just means that it
+ *                 has written early data to the server. To know whether the
+ *                 server has accepted early data or not, you should call
+ *                 mbedtls_ssl_get_early_data_status() with the handshake
+ *                 completed.
+ */
+int mbedtls_ssl_write_early_data( mbedtls_ssl_context *ssl,
+                                  const unsigned char *buf, size_t len );
+
+#define MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_SENT  0
```

Comment:
MBEDTLS_SSL_EARLY_DATA_STATUS_CAN_DO means SSL context allow early data, any negative value means not allow.
MBEDTLS_SSL_EARLY_DATA_STATUS_ALLOW_SEND, I am not sure if it should be expose to user. that means `mbedtls_ssl_write_early_data` can called without error report. If not expose that, we should call `mbedtls_ssl_handshake_step` inside `mbedtls_ssl_write_early_data`
```suggestion
#define MBEDTLS_SSL_EARLY_DATA_STATUS_CAN_DO  0
#define MBEDTLS_SSL_EARLY_DATA_STATUS_ALLOW_SEND 3
```

Reply:
The SSL context is set-up by the application and if it wants to send early data it has to set it up with a proper PSK, enabling early data with `mbedtls_ssl_tls13_conf_early_data()` ... Thus the application is aware if the SSL context potentially allow early data or not. That's why there is currently no API to tell it so.

Setting-up the SSL context is not trivial though thus we should at least have clear debug logs when early data have been enabled through `mbedtls_ssl_tls13_conf_early_data()` but eventually early data cannot be sent.

Reply:
One more question. Is below code expected ?
```
void mbedtls_ssl_tls13_conf_early_data( mbedtls_ssl_context *ssl, int enabled)
{
    ssl->conf->early_data_enabled = MBEDTLS_SSL_EARLY_DATA_DISABLED;
    if( enabled == MBEDTLS_SSL_EARLY_DATA_ENABLED)
    {
        if( ssl context can not do early data)
        {
            print(reason)
            return;
        }
        ssl->conf->early_data_enabled = MBEDTLS_SSL_EARLY_DATA_ENABLED;
    }
}

int mbedtls_ssl_get_early_data_status( mbedtls_ssl_context *ssl )
{
    if( is_server || handshake_is_not_over )
        return( BAD_INPUT_DATA );
    return( ssl->early_data_status );
}

```